### PR TITLE
feat: Run tests in local gems

### PR DESCRIPTION
**Pull request by bibendi**
🙈 

Linked with https://github.com/abak-press/pulscen/pull/14387

WIP

### CODE REVIEW

File: lib/psp/manager.rb [hunk 1]
```diff
@@ -127,8 +164,15 @@ def empty_project_runners_allocation
     end
 
     def separate(collection)
-      collection.reduce([[], []]) do |memo, resolver|
-        memo[resolver.project? ? 0 : 1] << resolver
+      collection.reduce([[], [], []]) do |memo, resolver|
```
Comment (apress-hound) [inline lib/psp/manager.rb]: Use each_with_object instead of reduce.

File: lib/psp/manager.rb [hunk 2]
```diff
@@ -112,13 +127,35 @@ def calculate_plugin_runners_allocation
       end
     end
 
+    def calculate_gem_runners_allocation
+      return empty_gem_runners_allocation if @gm_runners_count.zero?
+
+      allocation = @gems.flat_map(&:directory)
+                     .in_groups(@gm_runners_count, false)
+                     .reject(&:empty?)
```
Comment (apress-hound) [inline lib/psp/manager.rb]: Use 2 (not 15) spaces for indenting an expression in an assignment spanning multiple lines.

File: lib/psp/manager.rb [hunk 3]
```diff
@@ -112,13 +127,35 @@ def calculate_plugin_runners_allocation
       end
     end
 
+    def calculate_gem_runners_allocation
+      return empty_gem_runners_allocation if @gm_runners_count.zero?
+
+      allocation = @gems.flat_map(&:directory)
+                     .in_groups(@gm_runners_count, false)
```
Comment (apress-hound) [inline lib/psp/manager.rb]: Use 2 (not 15) spaces for indenting an expression in an assignment spanning multiple lines.

File: lib/psp/manager.rb [hunk 4]
```diff
@@ -74,12 +82,19 @@ def print_pl_runners_allocation
       (@pl_runners_alloc.map { |x| Ascii.magenta(x) } + empty) * '/'
     end
 
+    def print_gm_runners_allocation
+      empty = @gm_runners_count.zero? ? [Ascii.green('0')]
```
Comment (apress-hound) [inline lib/psp/manager.rb]: Avoid multi-line ternary operators, use if or unless instead.

File: lib/psp/manager.rb [hunk 5]
```diff
@@ -51,12 +57,14 @@ def print_runners_allocation
...
-       print_pl_runners_allocation] * ' + '
+       print_pl_runners_allocation,
+       print_gm_runners_allocation] * ' + '
     end
 
     def print_runners_count
       [Ascii.blue(@pr_runners_count),
-       Ascii.magenta(@pl_runners_count)] * ' + '
+       Ascii.magenta(@pl_runners_count),
+       Ascii.green(@gm_runners_count)] * ' + '
```
Comment (apress-hound) [inline lib/psp/manager.rb]: Favor Array#join over Array#*.

File: lib/psp/manager.rb [hunk 6]
```diff
@@ -51,12 +57,14 @@ def print_runners_allocation
       allocation unless defined?(@allocation)
 
       [print_pr_runners_allocation,
-       print_pl_runners_allocation] * ' + '
+       print_pl_runners_allocation,
+       print_gm_runners_allocation] * ' + '
```
Comment (apress-hound) [inline lib/psp/manager.rb]: Favor Array#join over Array#*.

File: lib/psp/manager.rb [hunk 7]
```diff
@@ -19,20 +20,21 @@ def initialize(collection, options = Hash.new)
...
-      @project, @plugins = separate(collection)
+      @project, @plugins, @gems = separate(collection)
 
       # TODO : Напрашивается 2-а класса проектные раннеры и плагинные
-      @pr_runners_count, @pl_runners_count = calculate_runners_count
-      @pr_runners_alloc, @pl_runners_alloc = Array.new, Array.new
-      @pr_runners_satiety, @pl_runners_satiety = 0, 0
+      @pr_runners_count, @pl_runners_count, @gm_runners_count = calculate_runners_count
+      @pr_runners_alloc, @pl_runners_alloc, @gm_runners_alloc = Array.new, Array.new, Array.new
+      @pr_runners_satiety, @pl_runners_satiety, @gm_runners_satiety = 0, 0, 0
```
Comment (apress-hound) [inline lib/psp/manager.rb]: Do not use parallel assignment.

File: lib/psp/manager.rb [hunk 8]
```diff
@@ -19,20 +20,21 @@ def initialize(collection, options = Hash.new)
...
 
-      @project, @plugins = separate(collection)
+      @project, @plugins, @gems = separate(collection)
 
       # TODO : Напрашивается 2-а класса проектные раннеры и плагинные
-      @pr_runners_count, @pl_runners_count = calculate_runners_count
-      @pr_runners_alloc, @pl_runners_alloc = Array.new, Array.new
-      @pr_runners_satiety, @pl_runners_satiety = 0, 0
+      @pr_runners_count, @pl_runners_count, @gm_runners_count = calculate_runners_count
+      @pr_runners_alloc, @pl_runners_alloc, @gm_runners_alloc = Array.new, Array.new, Array.new
```
Comment (apress-hound) [inline lib/psp/manager.rb]: Do not use parallel assignment.

File: lib/psp/file_resolver.rb [hunk 9]
```diff
@@ -38,5 +42,9 @@ def relative_path(full_path)
     def extract_plugin_name(path)
       path.match(/vendor\/plugins\/(?<name>[\w\_\-]+)\/spec$/)[:name]
     end
+
+    def extract_gem_name(path)
+      path.match(/vendor\/gems\/(?<name>[\w\_\-]+)\/spec$/)[:name]
```
Comment (apress-hound) [inline lib/psp/file_resolver.rb]: Use %r around regular expression.

File: lib/psp/manager.rb [hunk 10]
```diff
@@ -137,12 +181,15 @@ def calculate_runners_count
       if @project.empty?
         [0, @concurrency]
       elsif @plugins.empty?
-        [@concurrency, 0] # TODO : Добавить маленькие очереди и сюда!
+        [@concurrency, 0]
+      elsif @gems.empty?
+        [@concurrency, 0]
       else
```
Comment (Napolskih) [inline lib/psp/manager.rb]: эти кейсы конечно не работают, но тут походу надо 3 элемента возвращать

**Comment by bibendi** [comment]
все оказалось проще https://github.com/abak-press/psp/pull/17

**Pull request closed without merging**